### PR TITLE
Issue 57.1

### PR DESCRIPTION
**Pull request by sanjuthomas**
Issue 57 code changes -
1. JsonItemWriter  - New class
2. JsonItemProcessor - New class
3. LoadDocumentsFromDirectoryJob - Changed class to deal with JSON input also
4. URIs generated using default strategy, I asked this question earlier and the answer was to load the json as it is. So the URI is the absolute file location of the JSON.
5. LoadJsonDocumentsFromDirectoryJobTest - Test case to load the JSON, we add assertion to load and compare file when we have a URI generation/creation strategy.

### CODE REVIEW

**Comment by sastafford** [comment]
@sanjuthomas - I ran the LoadJsonDocumentsFromDirectoryJobTest on my local machine.  I'm running it under Windows and I got the following REST API error: 

Bad Request. Server Message: REST-INVALIDPARAM: (err:FOER0000) Invalid parameter: invalid uri: E:\world-bank\marklogic-spring-batch\build\resources\test\data\elmo.json

Watching the URI variable in the JsonItemWriter on line 28, the uri value is as follows: 
URI = "E:\\world-bank\\marklogic-spring-batch\\build\\resources\\test\\data\\elmo.json"

Probably doesn't like the double slashes.

File: src/test/java/com/marklogic/spring/batch/job/LoadJsonDocumentsFromDirectoryJobTest.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import org.springframework.mock.env.MockPropertySource;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.support.AnnotationConfigContextLoader;
+
+@ContextConfiguration(classes = {
+        com.marklogic.junit.spring.BasicTestConfig.class,
+        LoadDocumentsFromDirectoryJob.class,
+        com.marklogic.spring.batch.test.MarkLogicSpringBatchTestConfig.class,
+        com.marklogic.spring.batch.item.JsonItemWriter.class,
+        com.marklogic.spring.batch.item.JsonItemProcessor.class
```
Comment (sastafford) [inline src/test/java/com/marklogic/spring/batch/job/LoadJsonDocumentsFromDirectoryJobTest.java]: The test class should not really have any idea of the ItemWriter or ItemProcessor being used.  Try adding a conditional statement in the LoadDocumentsFromDirectoryJob to instantiate the proper ItemWriter / ItemProcessor

**Comment by sastafford** [comment]
Apply this concept to the ItemWriter and ItemProcessor in the LoadDocumentsFromDirectoryJob class.  Does this make sense?

```
    @Bean
    public ItemWriter<Document> writer() {
        if("JSON".equalsIgnoreCase(documentType)) {
            return new JsonItemWriter();
        } else {
            return new DocumentItemWriter();
        }

    }
```

**Comment by sastafford** [comment]
This is an example of what it might look like in the LoadDocumentsFromDirectoryJob class.

The bean that gets used would be based on some sort of condition of the document_type.  This is just a suggestion. 

```Java
    @Conditional(xxx)
    @Bean
    public ItemWriter<Document> xmlWriter() {
        return new DocumentItemWriter();
    }

    @Conditional(xxx)
    @Bean
    public ItemWriter<ObjectNode> jsonWriter() {
        return new JsonItemWriter();
    }
```

**Comment by sanjuthomas** [comment]
@sastafford Yes, the URI is an issues. we have to have a URI strategy. In this case, I used the default option which was available there.

**Comment by sastafford** [comment]
My default answer is to adapt mlcp uri strategy.  I have another issue for uri generation but have not started.

 
 
  On Tue, May 24, 2016 at 5:31 PM, Sanju Thomas<notifications@github.com> wrote:   
@sastafford Yes, As I mentioned in the pull request I used the absolute path of the file as URI. Slash could be an issue in Windows. We have to have a URI strategy. In this case, I used the default option which was available there. I remember, you mentioned that the URI generation is part of another story.

**Comment by sanjuthomas** [comment]
@sastafford  OK, went ahead with the URI generation for this one. Conditional wiring is also, let me know.

File: src/main/java/com/marklogic/spring/batch/item/JsonItemProcessor.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.core.io.Resource;
+
+/**
+ * Created by sanjuthomas on 5/24/16.
+ */
+public class JsonItemProcessor implements ItemProcessor<Resource, ObjectNode> {
+
+    private static final String URI = "uri";
+
+    private static final ObjectMapper MAPPER = new ObjectMapper();
```
Comment (rjrudin) [inline src/main/java/com/marklogic/spring/batch/item/JsonItemProcessor.java]: Thoughts on making this non-static, non-final? Then a client can use this class and optionally provider their own ObjectMapper.

File: src/main/java/com/marklogic/spring/batch/item/JsonItemProcessor.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import com.marklogic.uri.DefaultUriGenerator;
+import org.springframework.batch.item.ItemProcessor;
+import org.springframework.core.io.Resource;
+
+/**
+ * Created by sanjuthomas on 5/24/16.
+ */
+public class JsonItemProcessor implements ItemProcessor<Resource, ObjectNode> {
+
+    private static final String URI = "uri";
```
Comment (rjrudin) [inline src/main/java/com/marklogic/spring/batch/item/JsonItemProcessor.java]: This looks unused?

File: src/main/java/com/marklogic/spring/batch/item/JsonItemWriter.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    private UriGenerator uriGenerator = new DefaultUriGenerator();
+
+    private static final String URI = "uri";
+
+    @Autowired
+    private DatabaseClientProvider databaseClientProvider;
+
+    @Override
+    public void write(List<? extends ObjectNode> items) throws Exception {
+        DatabaseClient client = databaseClientProvider.getDatabaseClient();
```
Comment (rjrudin) [inline src/main/java/com/marklogic/spring/batch/item/JsonItemWriter.java]: I recommend having this class take a DatabaseClient as a constructor argument. The DatabaseClientProvider interface provides a level of indirection that isn't really needed here. All this class cares about is being given an instance of DatabaseClient.

File: src/main/java/com/marklogic/spring/batch/item/JsonItemWriter.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.List;
+
+/**
+ * Created by sanjuthomas on 5/24/16.
+ */
+public class JsonItemWriter implements ItemWriter<ObjectNode> {
+
+    private UriGenerator uriGenerator = new DefaultUriGenerator();
+
+    private static final String URI = "uri";
```
Comment (rjrudin) [inline src/main/java/com/marklogic/spring/batch/item/JsonItemWriter.java]: This looks unused?

File: src/main/java/com/marklogic/spring/batch/item/JsonItemProcessor.java (hunk 3 above)
Comment (sanjuthomas) [inline src/main/java/com/marklogic/spring/batch/item/JsonItemProcessor.java]: @rjrudin I removed that.

File: src/main/java/com/marklogic/spring/batch/item/JsonItemWriter.java (hunk 4 above)
Comment (sanjuthomas) [inline src/main/java/com/marklogic/spring/batch/item/JsonItemWriter.java]: @rjrudin  I think that make sense, will do that.

File: src/main/java/com/marklogic/spring/batch/item/JsonItemWriter.java (hunk 5 above)
Comment (sanjuthomas) [inline src/main/java/com/marklogic/spring/batch/item/JsonItemWriter.java]: @rjrudin thats removed.

File: src/test/java/com/marklogic/spring/batch/job/LoadJsonDocumentsFromDirectoryJobTest.java (hunk 1 above)
Comment (sanjuthomas) [inline src/test/java/com/marklogic/spring/batch/job/LoadJsonDocumentsFromDirectoryJobTest.java]: @sastafford  this is taken care.

File: src/main/java/com/marklogic/spring/batch/item/JsonItemProcessor.java (hunk 2 above)
Comment (sanjuthomas) [inline src/main/java/com/marklogic/spring/batch/item/JsonItemProcessor.java]: @rjrudin  not sure.

**Comment by sastafford** [comment]
Looks good.  Ran through all the tests and everything passed.  One minor change and I'll merge the changes

Can you rename this file name for this class.

https://github.com/sastafford/marklogic-spring-batch/blob/issue-57.1/src/main/java/com/marklogic/spring/batch/job/XMLDocumentTypeCondition.java


It should be XmlDocumentTypeCondition.java

**Comment by sanjuthomas** [comment]
@sastafford  interesting, its showing as small letters in my ide. even the class name small letter. humm, how did it compile then?

**Pull request merged**